### ISSUE TITLE
[POSIX] builtins `test` and `read`

### ISSUE THREAD

**Issue by d630** [hut/lscd#1]
Hi,

I have been trying to run `lscd` as script (with and without arguments), but I always get into the while loop with:

``` sh
var/tmp/lscd: 185: read: Illegal option -n
var/tmp/lscd: 188: read: Illegal option -s
var/tmp/lscd: 189: read: Illegal option -s
var/tmp/lscd: 190: read: Illegal option -s
var/tmp/lscd: 191: [: unexpected operator
^C
```

My standard shell is:

``` sh
$ readlink /bin/sh
dash
```

After reading this [site](http://mywiki.wooledge.org/Bashism), I recommend using test conditionals with single equals sign (=) and `read` only with option `-r` (because of POSIX compatibility).

**Comment by hut** [comment]
I was trying to avoid bashisms, but I'm afraid `read -n 1` is kind of necessary for this script to work, unless there is another way to extract the user input character by character.

**Comment by d630** [comment]
Ja, I see that.

Instead of `read -n 1`, I am thinking of doing:

``` sh
read input

# printf
input="$(printf '%c' "$input")"

# cut
input="$(echo "$input" | cut -c 1)"

# expr
input="$(expr "x$input" : "x.\{,0\}\(.\{,1\}\)")"
```

Instead `read -e` of using `rlwrap` or the like. Maybe, you could read `k{1..3}` with the help of a while loop?!

**Comment by hut** [comment]
The problem is rather that a plain "read" waits for enter so you can't use single-key bindings.

I tried it using "stty raw", check out the "noread" branch if it suits your needs. It's not perfect yet, for example, you can't use arrow keys and so on.

**Comment by d630** [comment]
Ah, ok.

Based on [L1](http://www.etalabs.net/sh_tricks.html) and [L2](http://www.unix.com/shell-programming-and-scripting/110380-using-arrow-keys-shell-scripts.html), one could do something like that with every arrow key:

``` sh
# up arrow key
read _ cuu1 <<-EOF
$(tput cuu1 2>&1 | od -t o1)
EOF
```

And then:

``` sh
# getc
stty raw
read _ input <<-EOF
$(dd bs=10 count=1 status=none | od -t o1)
EOF
stty cooked

# select
case "$input" in
    k|"$cuu1")
            move -1
            ;;
    q)
            on_exit
            break
            ;;
esac
```

**Comment by hut** [comment]
That's a neat idea. I didn't integrate it yet, but nevertheless, I merged the noread branch into the master branch now. Can you start and operate the program now, using h/j/k/l instead of arrow keys?

**Comment by d630** [comment]
1. You have forgotten to edit the equal signs: `sed -i 's/==/=/g' ~/var/tmp/lscd`
2. Key `l` is not working because of `stat` in `openfile()`. It is like with `ls` and other coreutils: the format flag respects my language settings. In my case, it would be working with:

``` sh
        case "$filetype" in
           directory|Verzeichnis)
                movedir "$f";;
            *)
                $STRING_file_opener "$f";;
        esac
```

I recommend to use `file` instead:

``` sh
filetype="$(file --mime-type -bL "$f")"
```

or:

``` sh
test -d
```

**Comment by d630** [comment]
He,

I have forked this project and started to [rewrite](https://github.com/D630/gist-bin-pub/tree/master/blscd) it in `bash`. Different is:
- view with four columns
- using `find` instead of `ls`
- the header is an `ls` output and shows also the position of the cursor
- using `tput` instead of ascii code
- no colorizing of the listings
- using the alternate screen of the terminal where to start the application

See, also [deer](https://github.com/Vifon/deer), a ranger like navigator in `zsh`

**Comment by hut** [comment]
Great to hear!

Do you just do it for fun or because you find it useful? I would probably never trade ranger for lscd unless I have to. I work as a sysadmin on some computers where the few dependencies of ranger are not available.

**Comment by d630** [comment]
Just for fun. I love checking things out with my shell resp. `bash`.

I like `ranger` in its entirety, particularly its navigation, but its start time is a bit too long for me. I am accustomed to work with several temporarily running `xterm` windows/instances, not to multiplexers etc.

**Comment by hut** [comment]
Yeah, I've been attempting to rewrite ranger a couple of times in different languages to improve the efficiency, and the start-up time in particular, but I always fall back to the python version because I've invested so much time on it. But the shellscript version is just ridiculously funny. Especially how you managed to throw in miller columns :)

**Comment by d630** [comment]
Ja, shell scripting seems to be nothing more than string manipulation ;)

The term "Miller columns" was new to me. Thanks

**Comment by d630** [comment]
He Hut, i have moved my fork and released a first one as [v0.1.0.0](https://github.com/D630/blscd). Have a look at it
